resources/files/windows/PuppetBolt/PuppetBoltFunctions.ps1

```diff
@@ -0,0 +1,1639 @@
+    * For Windows nodes, specify the winrm:// protocol if it has not be configured
+    * For SSH, port defaults to `22`
+    * For WinRM, port defaults to `5985` or `5986` based on the --[no-]ssl setting
+    .PARAMETER Query
+    [One of -Targets or -Query is required] Query PuppetDB to determine the targets
+    .PARAMETER Parameters
+    Parameters to a task or plan as:
+    * a valid json string
+    * powershell HashTable
+    * a json file: '@<file>'
+    * on stdin: '-'
+    .PARAMETER Rerun
+    Retry on nodes from the last run
+    * 'all' all nodes that were part of the last run.
+    * 'failure' nodes that failed in the last run.
+    * 'success' nodes that succeeded in the last run.
+    .PARAMETER Noop
+    Execute a task that supports it in noop mode
+    .PARAMETER Description
+    Description to use for the job
+    .PARAMETER User
+    User to authenticate as
+    .PARAMETER Password
+    Password to authenticate with. Omit the value to prompt for the password.
+    .PARAMETER SSHPrivateKey
+    Private ssh key to authenticate with
+    .PARAMETER HostKeyCheck
+    Check host keys with SSH
+    .PARAMETER SSL
+    Use SSL with WinRM
+    .PARAMETER SSLVerify
+    Verify remote host SSL certificate with WinRM
+    .PARAMETER RunAs
+    User to run as using privilege escalation
+    .PARAMETER SudoPassword
+    Password for privilege escalation. Omit the value to prompt for the password.
+    .PARAMETER Concurrency
+    Maximum number of simultaneous connections (default: 100)
+    .PARAMETER CompileConcurrency
+    Maximum number of simultaneous manifest block compiles (default: number of cores)
+    .PARAMETER ModulePath
+    List of directories containing modules, separated by ';'
+    .PARAMETER BoltDir
+    Specify what Boltdir to load config from (default: autodiscovered from current working dir)
+    .PARAMETER ConfigFile
+    Specify where to load config from (default: ~/.puppetlabs/bolt/bolt.yaml)
+    .PARAMETER InventoryFile
+    Specify where to load inventory from (default: ~/.puppetlabs/bolt/inventory.yaml)
+    .PARAMETER SaveRerun
+    Whether to update the rerun file after this command.
+    .PARAMETER Transport
+    Specify a default transport: ssh, winrm, pcp, local, docker, remote
+    .PARAMETER ConnectionTimeout
+    Connection timeout (defaults vary)
+    .PARAMETER TTY
+    Request a pseudo TTY on targets that support it
+    .PARAMETER Tmpdir
+    The directory to upload and execute temporary files on the target
+    .PARAMETER Color
+    Whether to show output in color
+    .PARAMETER OutputLevel
+    Level of output logging to display (defaults to no output). Must be one of:
+    * 'Verbose' only verbose logging
+    * 'Debug' both verbose and debug logging
+    * 'Trace' all of verbose, debug, and stacktrace logging
+    #>
+    [CmdletBinding()]
+    param(
+        [parameter(Mandatory,ValueFromPipeline=$True,ValueFromPipelineByPropertyName,Position=0)]
```

Comment:
Surely some of these parameters are mandatory and should be marked as such?

Any thoughts to using parameter sets to be more intuitive for PowerShell users instead of smashing all parameters here?